rust/operator-binary/src/discovery.rs

```diff
@@ -0,0 +1,160 @@
+) -> Result<Vec<ConfigMap>, Error> {
+    let name = owner.name();
+    Ok(vec![
+        build_discovery_configmap(&name, owner, hive, chroot, pod_hosts(hive)?)?,
+        // TODO: do we need that - i think there is only internal access required?
+        build_discovery_configmap(
+            &format!("{}-nodeport", name),
+            owner,
+            hive,
+            chroot,
+            nodeport_hosts(client, svc, "hive").await?,
+        )?,
+    ])
+}
+
+/// Build a discovery [`ConfigMap`] containing information about how to connect to a certain [`HiveCluster`]
+///
+/// `hosts` will usually come from either [`pod_hosts`] or [`nodeport_hosts`].
+fn build_discovery_configmap(
+    name: &str,
+    owner: &impl Resource<DynamicType = ()>,
+    hive: &HiveCluster,
+    chroot: Option<&str>,
+    hosts: impl IntoIterator<Item = (impl Into<String>, u16)>,
+) -> Result<ConfigMap, Error> {
+    let mut conn_str = hosts
+        .into_iter()
+        .map(|(host, port)| format!("thrift://{}:{}", host.into(), port))
+        .collect::<Vec<_>>()
+        .join("\n");
+    if let Some(chroot) = chroot {
+        if !chroot.starts_with('/') {
+            return RelativeChroot { chroot }.fail();
+        }
+        conn_str.push_str(chroot);
+    }
+    Ok(ConfigMapBuilder::new()
+        .metadata(
+            ObjectMetaBuilder::new()
+                .name_and_namespace(hive)
+                .name(name)
+                .ownerreference_from_resource(owner, None, Some(true))
+                .with_context(|| ObjectMissingMetadataForOwnerRef {
+                    hive: ObjectRef::from_obj(hive),
+                })?
+                .with_recommended_labels(
+                    hive,
+                    APP_NAME,
+                    hive_version(hive).unwrap(),
+                    &HiveRole::MetaStore.to_string(),
+                    "discovery",
+                )
+                .build(),
+        )
+        .add_data("HIVE", conn_str)
+        .build()
+        .unwrap())
+}
+
+/// Lists all Pods FQDNs expected to host the [`HiveCluster`]
+fn pod_hosts(hive: &HiveCluster) -> Result<impl IntoIterator<Item = (String, u16)> + '_, Error> {
+    Ok(hive
+        .pods()
+        .context(ExpectedPods)?
+        .into_iter()
+        .map(|pod_ref| (pod_ref.fqdn(), APP_PORT)))
+}
+
+/// Lists all nodes currently hosting Pods participating in the [`Service`]
+async fn nodeport_hosts(
+    client: &stackable_operator::client::Client,
+    svc: &Service,
+    port_name: &str,
+) -> Result<impl IntoIterator<Item = (String, u16)>, Error> {
+    let svc_port = svc
+        .spec
+        .as_ref()
+        .and_then(|svc_spec| {
+            svc_spec
+                .ports
+                .as_ref()?
+                .iter()
+                .find(|port| port.name.as_deref() == Some("hive"))
+        })
+        .context(NoServicePort { port_name })?;
+    let node_port = svc_port.node_port.context(NoNodePort { port_name })?;
```

Comment:
Looks like ports are pretty much hard coded now (which is good) so this feels like overkill.

Reply:
Probably, id like too keep it in case we decide to have them configurable.